feat: add ignoreDaemonSetsUtilization and zeroOrMaxNodeScaling to NodeGroupAutoscalingOptions

jincong8973 opened the pull request:
Add missing field ignoreDaemonSetsUtilization and zeroOrMaxNodeScaling to externalgrpc NodeGroupAutoscalingOptions

#### What type of PR is this?

feature


#### What this PR does / why we need it:

externalgrpc GetOptions missed fields:
- ignoreDaemonSetsUtilization: we can't specific  it on NodeGroup Options or config `--ignore-daemonsets-utilization=true`
- zeroOrMaxNodeScaling: it's won't be useful in NodeGroup Options


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

dbonfigli commented:
/ok-to-test

dbonfigli commented:
thanks for this, it is indeed a while since we don't update the proto with the new options.

It looks all good, but if possible I would appreciate a ton if you could do this:
* since we are adding a missing option, let's also add the other one that is missing, `zeroOrMaxNodeScaling`, in the `NodeGroupAutoscalingOptions` message before `IgnoreDaemonSetsUtilization`;
* use lowercase for the first character of the filed `IgnoreDaemonSetsUtilization` in the message (I know `MaxNodeProvisionTime` is not respecting this, it was a mistake unfortunately);
* add the 2 new fields also in `externalgrpc_node_group_test.go` tests;
* add the 2 new fields also in `wrapper.go` example;
* bonus: update the readme.md "Code Generation" section adding that we need to run `go mod vendor` before generating the gRPC code, since now the CA is not vendorizing anymore the dependencies and without it imported proto files are not found.

Thanks again if you can also do this.

jincong8973 commented:
> thanks for this, it is indeed a while since we don't update the proto with the new options.
> 
> It looks all good, but if possible I would appreciate a ton if you could do this:
> 
> * since we are adding a missing option, let's also add the other one that is missing, `zeroOrMaxNodeScaling`, in the `NodeGroupAutoscalingOptions` message before `IgnoreDaemonSetsUtilization`;
> * use lowercase for the first character of the filed `IgnoreDaemonSetsUtilization` in the message (I know `MaxNodeProvisionTime` is not respecting this, it was a mistake unfortunately);
> * add the 2 new fields also in `externalgrpc_node_group_test.go` tests;
> * add the 2 new fields also in `wrapper.go` example;
> * bonus: update the readme.md "Code Generation" section adding that we need to run `go mod vendor` before generating the gRPC code, since now the CA is not vendorizing anymore the dependencies and without it imported proto files are not found.
> 
> Thanks again if you can also do this.


Thank you for the thorough review!  I've implemented all the requested changes:

- Added the missing zeroOrMaxNodeScaling field in NodeGroupAutoscalingOptions before ignoreDaemonSetsUtilization

- Lowercased the ignoreDaemonSetsUtilization field name to follow convention

Included both new fields in:

- `externalgrpc_node_group_test.go` tests, add a case for defaultOptions 

- `wrapper.go` example implementation

- Updated the "Code Generation" section in README.md to explicitly mention `go mod vendor `as a prerequisite

Please let me know if any additional adjustments are needed. very appreciate your guidance! @dbonfigli

dbonfigli commented:
awesome, please `gofmt` the files that are failing the formatting test, update the pr comment specifying that we are adding also the zeroOrMaxNodeScaling option and I think we are all set.

jincong8973 commented:
Everything should be addressed now. I have squashed related commits to streamline the history.

dbonfigli commented:
/approve

dbonfigli commented:
/lgmt

dbonfigli commented:
/lgtm
